### DISCUSSION TITLE
Best way to fetch release version from a GitHub release and insert it into a code-block

### DISCUSSION THREAD

**Discussion by Andre601** [mkdocs/mkdocs#3407] [Q&A]
I would like to set up a way to fetch the latest release info from a GitHub repository to then include within a Codeblock.

My first thought was to set up a JS file that would do this for me. Benefit would be that the info is up-to-date, meaning I wouldn't need to rebuild the docs after any update, but downside from what I can tell is, that the JS would be used on final page, meaning I would need to use query selectors and whatnot to find and replace the value with, or build the code-block myself, which could be tedious with the code-highlighting (I'm not too good in JS, so I don't know what options I have available here)?

My other idea was to add a python script that replaces a placeholder in Markdown's `on_page_markdown(...)` and add it as a hook.
Benefit here would be the use during the markdown processing, meaning I won't have to deal with code highlighting or HTML.
Downside is, that I have no idea how fetching data, processing it as JSON (Getting the values from it) and replacing a placeholder with it would work. Not to mention would I also want to cache the value to not exceed possibly low requests towards the GitHub API, which I also have no idea of how to do here...

So... What would be good solutions?

**Comment by oprypin** [comment]
If using GitHub Actions or relying on any other CI solution:
1. create a dummy file with example output that will not be updated.
   https://github.com/crystal-lang/crystal-book/blob/master/crystal-version.txt

2. reference it in the site using pymdownx.include:
   https://github.com/crystal-lang/crystal-book/blob/9a56ae58f73a55e4b48e6e1afcc7503ca3bc6aa0/docs/getting_started/README.md?plain=1#L17

3. Replace the file contents as a step in CI prior to building it, using any ability of GitHub Actions
   e.g. here we download the actual tool and run it to find its version
   https://github.com/crystal-lang/crystal-book/blob/9a56ae58f73a55e4b48e6e1afcc7503ca3bc6aa0/.github/workflows/deploy-book.yml#L46
   , but you can instead use any GitHub action like find-latest-release or https://github.com/oprypin/find-latest-tag to populate it 

The advantage of this solution is that the version will be fetched only in CI and will not slow down local builds
 The disadvantage is that this step is disjointed from the main build.
And, being a non-js solution, it is not refreshed real-time

**Reply by Andre601** [reply]
I don't use github actions tho...

**Reply by oprypin** [reply]
@Andre601 that's ok, in that case just find a different way to find out the latest release and write it into a file.

**Comment by kamilkrzyskow** [comment]
What exactly do you mean by "fetch GitHub release"? You want to get 1 file, or do you want to check the latest release and download a zip containing various source files? What is the repository if it's public? In response to Oprypin you said that you don't use GitHub Actions, so what CI are you using? I guess that it's Codeberg based on some other discussion thread, but to be clear I ask.

To have cache you need a backend of some sort, so the JavaScript method wouldn't have that assuming you'd directly access the GitHub API via GraphQL. I never done that, so I don't know what kind of rate limit there is or how to protect against spamming the API like you said. If you have some sort of backend server available you could fetch data through it and let it manage accessing and caching of the GitHub API call / result.
EDIT: I guess there is the option of storing fetched data in the Local Storage of the user in the browser, to avoid the need for a backend sever but I guess that I never seen such kind of use case.

So perhaps the best approach is to scope the API access to the CI or MkDocs build process. You could setup cache and/or setup a time based cron workflow (assuming that's available in your CI) to assure being somewhat up-to-date with the latest release via frequent rebuilds.

As for the methodology / approach of how to set it up in the end, there is really no single best way, I think. 
Without more details of what files you're getting, and what roadblocks you have I would say that the issue is generic enough that any boilerplate code from ChatGPT would be enough to guide you through the process 🤔

**Reply by Andre601** [reply]
The goal is to just fetch the latest release, namely the tag name of the release from a GitHub Repository.

Also, yes it's codeberg and the CI is Woodpecker-CI, meaning that depending on what I want to use I have to utilize certain images (for example, `curl` isn't part of the CI itself and I need to use a specific image to utilize it).

**Reply by kamilkrzyskow** [reply]
If there is no file related processing related to the fetch, then I agree with pawamoy below to use JavaScript, as the Material theme practically does the same thing with the repository heading and star amount banner, and there is no problem with rate limits etc.

As for the Woodpecker-CI, I've looked into it for a bit, and it does support any Docker images from the registry.
So I guess you can use `image: alpine:latest` and run the command `apk add curl` to add curl, and many other features, I do use it like that in GitLab CI.

**Comment by pawamoy** [comment]
I think you should go with Javascript. ChatGPT could easily give you a working snippet to fetch the name of the latest release of a project on GitHub, using its API. It's probably a single HTTP request, so I wouldn't worry about rate limits. If you somehow reach a limit, you can add a personal access token to authenticate your request.

**Reply by Andre601** [reply]
I don't use ChatGPT, nor will I ever want to touch this waste of energy and resources. Especially given its track-record of suggesting code that doesn't even work for the coding language the user asked help for.

**Reply by oprypin** [reply]
@Andre601 I hope to write up an answer here but I just need to find time for it in next days

**Reply by pawamoy** [reply]
Get JSON in plain JavaScript:

https://stackoverflow.com/a/35970894

Get latest GitHub release:

https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#get-the-latest-release

Maybe checkout how Material for MkDocs does it.

Then use a regex replace on the inner HTML of your code block.

https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementById

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace

**Reply by Andre601** [reply]
> Maybe checkout how Material for MkDocs does it.

[Already know how](https://codeberg.org/Andre601/AdvancedServerList/src/branch/master/docs/assets/js/repo-stats.js)

It's just that I'm not too experienced with JS so I didn't know of possible solutions to use here.

**Reply by kamilkrzyskow** [reply]
Here is a quick code snippet I run on https://asl.andre601.ch/examples/ via the Developer Tools.
It changes the contents of the first code block to the latest tag name of the MkDocs repository.
You can hook it up the same way as your other subscribe hook. 
You mentioned something about code highlighting, I don't get what there is to code highlight when it comes to a tag name, so I left that out.

```js
replaceCodeBlock = async () => {
  const url = 'https://api.github.com/repos/mkdocs/mkdocs/tags';
  const tags = await fetch(url).then(_ => _.json());
  const codeBlock = document.querySelector('.md-content pre code');
  const spanWrapper = document.createElement("span");
  spanWrapper.innerText = tags[0]['name'];
  codeBlock.replaceChildren(spanWrapper);
}

replaceCodeBlock()
```

**Reply by Andre601** [reply]
Wouldn't really do what I want...

I effectively want to have this
```groovy
repositorories {
    maven { url = 'https://jitpack.io/' }
}

dependencies {
    implementation 'ch.andre601.asl-api:api:{version}'

    // Optional platform dependencies
    implementation 'ch.andre601.asl-api:platform-bukkit:{version}'
    implementation 'ch.andre601.asl-api:platform-bungeecord:{version}'
    implementation 'ch.andre601.asl-api:platform-velocity:{version}'
}
```
...turn into this
```groovy
repositorories {
    maven { url = 'https://jitpack.io/' }
}

dependencies {
    implementation 'ch.andre601.asl-api:api:v3.1.0'

    // Optional platform dependencies
    implementation 'ch.andre601.asl-api:platform-bukkit:v3.1.0'
    implementation 'ch.andre601.asl-api:platform-bungeecord:v3.1.0'
    implementation 'ch.andre601.asl-api:platform-velocity:v3.1.0'
}
```
Your example would replace the entire content of the code block, while also only affecting the first one it finds when in my case, there would be 2 to change (In different content tabs).

Still apreciate the help you gave me. :)

**Reply by kamilkrzyskow** [reply]
> Wouldn't really do what I want...

Well of course, I would be quite surprised if a simple example would handle all of your edge cases. 

Looking at the new requirements, it looks like the code block on https://asl.andre601.ch/api/
Here is another code snippet, it's scoped to only work on pages with `/api/` in the URL.
It goes over every code block and replaces every `{version}` with the tag name retrieved from the GitHub API.
However, it looks to me as if the "fetch GitHub release" part is incorrect, because you want to fetch the data from the codeberg repository and I don't know what API is there for that 🤔 

```js
replaceCodeBlock = async () => {

  if (!location.href.includes('/api/')) {
    return;
  }

  const url = 'https://api.github.com/repos/mkdocs/mkdocs/tags';
  const tags = await fetch(url).then(_ => _.json());
  const version = tags[0]['name'];
  const versionToken = '{version}';
  const codeBlocks = document.querySelectorAll('.md-content pre code');
  for (const codeBlock of codeBlocks) {
    codeBlock.innerHTML = codeBlock.innerHTML.replace(new RegExp(versionToken, 'g'), version);
  }
}

replaceCodeBlock();
```

**Reply by Andre601** [reply]
The API is on GitHub so using GitHub here is actually correct.

**Reply by kamilkrzyskow** [reply]
OK, then you should be good to go with the last example.
